File: server/hub.js
#!/usr/bin/env node
'use strict';

/*
 * RBY MMO hub.
 *
 * A relay, not a game server. It never simulates anything: it owns who is
 * connected, where they said they are, and which two players are currently
 * paired for a trade or a battle. Everything else -- the trade state
 * machine, the lockstep battle -- runs inside the two game clients, exactly
 * as it does over a direct link, with this process forwarding the bytes.
 *
 * That split is deliberate. The engine's link code is the authority on Gen
 * 1 behaviour and it already works; a hub that tried to referee battles
 * would be a second, worse implementation of it that could disagree with
 * the clients.
 *
 * Wire format is newline-delimited JSON, which is what src/link/Net.lua's
 * relay backend already speaks, so the client reuses the engine's framing
 * instead of shipping its own.
 *
 * No dependencies: node hub.js [port]
 */

const net = require('net');

const PROTOCOL = 1;
const PORT = Number(process.argv[2] || process.env.RBY_MMO_PORT || 7788);
const HOST = process.env.RBY_MMO_HOST || '0.0.0.0';

const LOCAL_RADIUS = 12;
const NAME_MAX = 10;
const MESSAGE_MAX = 60;
const MAX_LINE = 64 * 1024;
const TIMEOUT_MS = 45000;
// The player cap. Defaults to 4 and may be raised to 64 -- the same bounds
// the in-game host picks from, and clamped here for the same reason they
// are clamped there: the limit must not be settable out of range by editing
// a config, only by the one mechanism that is meant to set it.
//
// A dedicated hub takes its limit from the environment. A hub running
// inside somebody's game takes it from the host's in-game choice; the two
// are the same protocol and the same bounds, just different front doors.
const MIN_PLAYERS = 2;
const MAX_PLAYERS = 64;
const DEFAULT_PLAYERS = 4;

function clampPlayers(value) {
  const n = Math.floor(Number(value));
  if (!Number.isFinite(n)) return DEFAULT_PLAYERS;
  return Math.min(MAX_PLAYERS, Math.max(MIN_PLAYERS, n));
}

const MAX_CLIENTS = process.env.RBY_MMO_MAX
  ? clampPlayers(process.env.RBY_MMO_MAX)
  : DEFAULT_PLAYERS;

/** id -> client */
const clients = new Map();
/** sessionId -> { a, b, kind } */
const sessions = new Map();

let nextId = 1;
let nextSession = 1;

const log = (...args) => console.log(new Date().toISOString(), ...args);

// ---------------------------------------------------------------- sanitising
//
// Everything below the line is untrusted: a client is somebody else's
// process, and a modified one is a normal thing to encounter. Every field
// is re-derived rather than checked in place.

const TEXT_OK = /[^A-Za-z0-9 .,!?'\-:;()/]/g;

function cleanText(value, limit) {
  if (typeof value !== 'string') return null;
  const clean = value.replace(TEXT_OK, '').replace(/\s+/g, ' ').trim();
  if (!clean) return null;
  return clean.slice(0, limit);
}

function cleanId(value) {
  if (typeof value !== 'string') return null;
  return /^[\w-]{1,40}$/.test(value) ? value : null;
}

// A sprite id is an engine identifier (SPRITE_RED), not prose. cleanText
// strips the underscore, turning it into SPRITERED, which then misses the
// catalog lookup and draws every player as the fallback -- invisibly,
// because the fallback works.
function cleanSpriteId(value) {
  if (typeof value !== 'string') return null;
  return /^\w{1,40}$/.test(value) ? value : null;
}

function cleanMapId(value) {
  if (typeof value !== 'string') return null;
  return /^[\w.-]{1,64}$/.test(value) ? value : null;
}

function cleanInt(value, min, max) {
  const n = Number(value);
  if (!Number.isFinite(n)) return null;
  const i = Math.floor(n);
  return i >= min && i <= max ? i : null;
}

const FACINGS = new Set(['up', 'down', 'left', 'right']);
const KINDS = new Set(['trade', 'battle']);
const SCOPES = new Set(['global', 'local', 'private']);

// ------------------------------------------------------------------ plumbing

function send(client, type, payload) {
  if (!client || client.socket.destroyed) return;
  const msg = Object.assign({}, payload, { type });
  // stringify can throw (RangeError on deeply nested input, TypeError on a
  // cycle). One peer's message must never be able to take the hub down for
  // everyone, so a message that will not serialise costs that connection
  // and nothing else.
  let line;
  try {
    line = JSON.stringify(msg) + '\n';
  } catch (err) {
    log(`dropping ${client.id}: message would not serialise (${err.message})`);
    client.socket.destroy();
    return;
  }
  client.socket.write(line);
}

// Relay payloads are forwarded unread, so shape is all that can be judged.
// Iterative on purpose: a recursive check would blow the same stack it is
// meant to protect. Mirrors Wire.payloadOk on the Lua side.
const PAYLOAD_MAX_DEPTH = 32;
const PAYLOAD_MAX_NODES = 4096;

function payloadOk(value) {
  if (value === null || typeof value !== 'object') return false;
  const stack = [[value, 1]];
  let nodes = 0;
  while (stack.length) {
    const [node, depth] = stack.pop();
    if (depth > PAYLOAD_MAX_DEPTH) return false;
    for (const child of Object.values(node)) {
      if (++nodes > PAYLOAD_MAX_NODES) return false;
      if (child !== null && typeof child === 'object') {
        stack.push([child, depth + 1]);
      }
    }
  }
  return true;
}

function fail(client, message) {
  send(client, 'mmo.error', { message });
  client.socket.end();
}

function presenceOf(client) {
  return {
    id: client.id,
    name: client.name,
    sprite: client.sprite,
    map: client.map,
    x: client.x,
    y: client.y,
    facing: client.facing,
    busy: Boolean(client.sessionId),
  };
}

function broadcast(type, payload, exceptId) {
  for (const client of clients.values()) {
    if (client.id !== exceptId && client.ready) send(client, type, payload);
  }
}

// ------------------------------------------------------------------ sessions

function peerOf(client) {
  if (!client.sessionId) return null;
  const session = sessions.get(client.sessionId);
  if (!session) return null;
  return clients.get(session.a === client.id ? session.b : session.a) || null;
}

function endSession(client, reason) {
  const id = client.sessionId;
  if (!id) return;
  const session = sessions.get(id);
  sessions.delete(id);
  client.sessionId = null;

  if (session) {
    const otherId = session.a === client.id ? session.b : session.a;
    const other = clients.get(otherId);
    if (other && other.sessionId === id) {
      other.sessionId = null;
      send(other, 'mmo.session_end', { reason });
      broadcast('mmo.move', presenceOf(other), other.id);
    }
  }
  if (clients.has(client.id)) broadcast('mmo.move', presenceOf(client), client.id);
}

function startSession(a, b, kind) {
  const id = String(nextSession++);
  sessions.set(id, { a: a.id, b: b.id, kind });
  a.sessionId = id;
  b.sessionId = id;

  // The requester hosts. Someone has to deal the battle's shared RNG seed,
  // and picking the side that asked keeps it deterministic rather than
  // racing on who answers first.
  send(a, 'mmo.session',
    { peer: b.id, peerName: b.name, kind, role: 'host', id });
  send(b, 'mmo.session',
    { peer: a.id, peerName: a.name, kind, role: 'guest', id });

  broadcast('mmo.move', presenceOf(a), a.id);
  broadcast('mmo.move', presenceOf(b), b.id);
  log(`session ${id}: ${a.name} <-> ${b.name} (${kind})`);
}

// ------------------------------------------------------------------ handlers

const handlers = {};

handlers['mmo.hello'] = (client, msg) => {
  if (client.ready) return;
  if (cleanInt(msg.proto, 0, 9999) !== PROTOCOL) {
    return fail(client,
      `This hub speaks protocol ${PROTOCOL}; your mod speaks ${msg.proto}.`);
  }
  const name = cleanText(msg.name, NAME_MAX);
  if (!name) return fail(client, 'That trainer name cannot be used here.');

  client.name = name;
  client.sprite = cleanSpriteId(msg.sprite) || 'SPRITE_RED';
  client.map = cleanMapId(msg.map);
  client.x = cleanInt(msg.x, 0, 4096);
  client.y = cleanInt(msg.y, 0, 4096);
  client.facing = FACINGS.has(msg.facing) ? msg.facing : 'down';
  client.ready = true;

  const players = [];
  for (const other of clients.values()) {
    if (other.ready && other.id !== client.id) players.push(presenceOf(other));
  }
  send(client, 'mmo.welcome', { id: client.id, players });
  broadcast('mmo.join', { player: presenceOf(client) }, client.id);
  log(`+ ${client.name} (${client.id}) -- ${clients.size} online`);
};

handlers['mmo.move'] = (client, msg) => {
  if (!client.ready) return;
  const map = cleanMapId(msg.map);
  const x = cleanInt(msg.x, 0, 4096);
  const y = cleanInt(msg.y, 0, 4096);
  if (map !== null && x !== null && y !== null) {
    client.map = map;
    client.x = x;
    client.y = y;
  } else {
    // no cell means "not in the world right now" (a battle, a menu): the
    // player stays on the roster but stops being placeable
    client.map = null;
    client.x = null;
    client.y = null;
  }
  if (FACINGS.has(msg.facing)) client.facing = msg.facing;
  broadcast('mmo.move', presenceOf(client), client.id);
};

handlers['mmo.chat'] = (client, msg) => {
  if (!client.ready) return;
  const scope = SCOPES.has(msg.scope) ? msg.scope : null;
  const text = cleanText(msg.text, MESSAGE_MAX);
  if (!scope || !text) return;

  const now = Date.now();
  // A light flood gate. Not a moderation system -- just enough that one
  // client cannot fill every other client's scrollback faster than it can
  // be read.
  if (now - client.lastChat < 500) return;
  client.lastChat = now;

  const payload = { from: client.id, name: client.name, scope, text };

  if (scope === 'private') {
    const target = clients.get(cleanId(msg.to));
    if (target && target.ready) send(target, 'mmo.chat', payload);
    return;
  }

  if (scope === 'local') {
    if (!client.map) return;
    for (const other of clients.values()) {
      if (other.id === client.id || !other.ready) continue;
      if (other.map !== client.map) continue;
      const distance = Math.max(
        Math.abs(other.x - client.x), Math.abs(other.y - client.y));
      if (distance <= LOCAL_RADIUS) send(other, 'mmo.chat', payload);
    }
    return;
  }

  broadcast('mmo.chat', payload, client.id);
};

handlers['mmo.request'] = (client, msg) => {
  if (!client.ready || client.sessionId) return;
  const kind = KINDS.has(msg.kind) ? msg.kind : null;
  const target = clients.get(cleanId(msg.to));
  if (!kind || !target || !target.ready || target.id === client.id) return;

  if (target.sessionId) {
    return send(client, 'mmo.decline', { name: target.name, kind });
  }
  client.pendingTo = target.id;
  send(target, 'mmo.request', { from: client.id, name: client.name, kind });
};

handlers['mmo.respond'] = (client, msg) => {
  if (!client.ready) return;
  const kind = KINDS.has(msg.kind) ? msg.kind : null;
  const asker = clients.get(cleanId(msg.to));
  if (!kind || !asker || !asker.ready) return;

  // only the player who was actually asked can answer, and only while the
  // ask is still outstanding
  if (asker.pendingTo !== client.id) return;
  asker.pendingTo = null;

  if (!msg.accept) {
    return send(asker, 'mmo.decline', { name: client.name, kind });
  }
  if (client.sessionId || asker.sessionId) {
    return send(asker, 'mmo.decline', { name: client.name, kind });
  }
  startSession(asker, client, kind);
};

handlers['mmo.relay'] = (client, msg) => {
  if (!client.ready || !client.sessionId) return;
  const peer = peerOf(client);
  if (!peer) return;
  if (cleanId(msg.to) !== peer.id) return;
  if (!payloadOk(msg.payload)) return;
  // The hub does not read the payload. It is the engine's own link
  // vocabulary, and interpreting it here would couple this process to a
  // protocol the game already owns.
  send(peer, 'mmo.relay', { from: client.id, payload: msg.payload });
};

handlers['mmo.session_leave'] = (client) => {
  endSession(client, 'peer_left');
};

handlers['mmo.ping'] = (client) => {
  send(client, 'mmo.pong', {});
};

// -------------------------------------------------------------------- server

const server = net.createServer((socket) => {
  if (clients.size >= MAX_CLIENTS) {
    socket.write(JSON.stringify({
      type: 'mmo.error',
      message: `This hub is full (${MAX_CLIENTS} players).`,
    }) + '\n');
    return socket.end();
  }

  const client = {
    id: String(nextId++),
    socket,
    buffer: '',
    ready: false,
    name: null,
    sprite: 'SPRITE_RED',
    map: null, x: null, y: null, facing: 'down',
    sessionId: null,
    pendingTo: null,
    lastChat: 0,
  };
  clients.set(client.id, client);

  socket.setNoDelay(true);
  socket.setTimeout(TIMEOUT_MS);
  socket.setEncoding('utf8');

  socket.on('data', (chunk) => {
    client.buffer += chunk;
    if (client.buffer.length > MAX_LINE) {
      return fail(client, 'Message too long.');
    }
    let index;
    while ((index = client.buffer.indexOf('\n')) >= 0) {
      const line = client.buffer.slice(0, index);
      client.buffer = client.buffer.slice(index + 1);
      if (!line) continue;

      let msg;
      try {
        msg = JSON.parse(line);
      } catch (err) {
        continue; // a malformed line is dropped, not fatal
      }
      if (!msg || typeof msg.type !== 'string') continue;

      const handler = handlers[msg.type];
      if (!handler) continue;
      try {
        handler(client, msg);
      } catch (err) {
        log(`handler ${msg.type} failed for ${client.id}:`, err.message);
      }
    }
  });

  const drop = () => {
    if (!clients.has(client.id)) return;
    endSession(client, 'peer_left');
    clients.delete(client.id);
    for (const other of clients.values()) {
      if (other.pendingTo === client.id) other.pendingTo = null;
    }
    if (client.ready) {
      broadcast('mmo.part', { id: client.id }, client.id);
      log(`- ${client.name} (${client.id}) -- ${clients.size} online`);
    }
  };

  socket.on('timeout', () => socket.destroy());
  socket.on('error', () => drop());
  socket.on('close', drop);
});

server.on('error', (err) => {
  log('hub failed to start:', err.message);
  process.exit(1);
});

server.listen(PORT, HOST, () => {
  log(`RBY MMO hub listening on ${HOST}:${PORT} (protocol ${PROTOCOL})`);
});

for (const signal of ['SIGINT', 'SIGTERM']) {
  process.on(signal, () => {
    log('shutting down');
    for (const client of clients.values()) {
      send(client, 'mmo.error', { message: 'The hub is shutting down.' });
      client.socket.end();
    }
    server.close(() => process.exit(0));
    setTimeout(() => process.exit(0), 1000).unref();
  });
}

File: src/Hub.lua
-- The hub, as pure logic.
--
-- This is the same relay `server/hub.js` implements, ported to Lua so a
-- player can host from inside the game. It owns who is connected, where
-- they last said they were, which two players are paired, and the player
-- cap the host chose. It does not simulate anything: trade and battle run
-- inside the two clients on the engine's own link code, and `mmo.relay`
-- payloads pass through unread.
--
-- **No sockets appear anywhere below.** Everything talks to *peer handles*
-- -- any table answering `:send(msg)` and `:close()`. `HostServer` supplies
-- socket-backed ones; the host's own client supplies an in-process one; the
-- suite supplies fakes. That is what lets the cap, the scope routing and
-- the session pairing be tested under plain luajit, which has no luasocket
-- and no LOVE.
--
-- Everything arriving here is untrusted -- it comes from another player's
-- process, and a modified one is a normal thing to meet -- so every field
-- is re-derived through Wire before it is believed.

local need = ...
local Config = need("Config")
local Wire = need("Wire")

local M = {}
M.__index = M

function M.new(opts)
  opts = opts or {}
  return setmetatable({
    limit = Config.clampPlayers(opts.maxPlayers),
    clients = {},     -- id -> client (greeted or not)
    count = 0,        -- connections
    players = 0,      -- of those, the ones that have said hello
    sessions = {},    -- sessionId -> { a, b, kind }
    nextId = 1,
    nextSession = 1,
    clock = 0,
  }, M)
end

-- Full means no room for another *player*. A connection that has not said
-- hello is not a player, and must not be able to hold a seat.
function M:isFull()
  return self.players >= self.limit
end

function M:pendingCount()
  return self.count - self.players
end

-- ------- plumbing

local function send(client, msgType, payload)
  if not client or not client.peer then return end
  local msg = {}
  if type(payload) == "table" then
    for k, v in pairs(payload) do msg[k] = v end
  end
  msg.type = msgType
  client.peer:send(msg)
end

local function presenceOf(client)
  return {
    id = client.id,
    name = client.name,
    sprite = client.sprite,
    map = client.map,
    x = client.x,
    y = client.y,
    facing = client.facing,
    busy = client.sessionId ~= nil,
  }
end

function M:broadcast(msgType, payload, exceptId)
  for id, client in pairs(self.clients) do
    if id ~= exceptId and client.ready then send(client, msgType, payload) end
  end
end

function M:refuse(peer, message)
  peer:send({ type = Wire.ERROR, message = message })
  peer:close()
end

-- Refuse someone who has a client record, and drop it in the same breath.
-- A connection turned away at hello will never become a player, so leaving
-- it in the table would hold a pending slot until the reaper came round.
function M:refuseClient(client, message)
  self:refuse(client.peer, message)
  self:drop(client)
end

-- ------- connection lifecycle

-- A peer that got this far has a socket (or a loopback) but has not said
-- hello yet, so it is not a player and does not appear on anyone's roster.
--
-- The player cap is therefore NOT applied here -- that is checked at hello.
-- Charging it on connect meant a peer that connected and said nothing held
-- a seat forever, so four silent sockets could lock everyone out of a
-- four-player game. What is bounded here is the number of un-greeted
-- connections, and update() reaps the ones that never introduce themselves.
function M:accept(peer)
  if self:pendingCount() >= Config.MAX_PENDING then
    self:refuse(peer, "Too many connections. Try again.")
    return nil
  end
  local client = {
    since = self.clock,
    id = tostring(self.nextId),
    peer = peer,
    ready = false,
    name = nil,
    sprite = Config.DEFAULT_SPRITE,
    map = nil, x = nil, y = nil, facing = "down",
    sessionId = nil,
    pendingTo = nil,
    lastChat = -math.huge,
  }
  self.nextId = self.nextId + 1
  self.clients[client.id] = client
  self.count = self.count + 1
  return client
end

function M:drop(client)
  if not client or not self.clients[client.id] then return false end
  self:endSession(client, "peer_left")
  self.clients[client.id] = nil
  self.count = self.count - 1
  if client.ready then self.players = self.players - 1 end
  -- an outstanding request pointed at a player who just left would let the
  -- asker wait forever for an answer nobody can give
  for _, other in pairs(self.clients) do
    if other.pendingTo == client.id then other.pendingTo = nil end
  end
  if client.ready then
    self:broadcast(Wire.PART, { id = client.id }, client.id)
  end
  return true
end

-- ------- sessions

function M:peerOf(client)
  local session = client.sessionId and self.sessions[client.sessionId]
  if not session then return nil end
  local otherId = session.a == client.id and session.b or session.a
  return self.clients[otherId]
end

function M:endSession(client, reason)
  local id = client.sessionId
  if not id then return end
  local session = self.sessions[id]
  self.sessions[id] = nil
  client.sessionId = nil

  if session then
    local otherId = session.a == client.id and session.b or session.a
    local other = self.clients[otherId]
    if other and other.sessionId == id then
      other.sessionId = nil
      send(other, Wire.SESSION_END, { reason = reason })
      self:broadcast(Wire.MOVE, presenceOf(other), other.id)
    end
  end
  if self.clients[client.id] then
    self:broadcast(Wire.MOVE, presenceOf(client), client.id)
  end
end

function M:startSession(a, b, kind)
  local id = tostring(self.nextSession)
  self.nextSession = self.nextSession + 1
  self.sessions[id] = { a = a.id, b = b.id, kind = kind }
  a.sessionId, b.sessionId = id, id

  -- The requester hosts. Someone has to deal the battle's shared RNG seed,
  -- and picking the side that asked keeps it deterministic rather than
  -- racing on who answers first.
  send(a, Wire.SESSION,
    { peer = b.id, peerName = b.name, kind = kind, role = "host", id = id })
  send(b, Wire.SESSION,
    { peer = a.id, peerName = a.name, kind = kind, role = "guest", id = id })

  self:broadcast(Wire.MOVE, presenceOf(a), a.id)
  self:broadcast(Wire.MOVE, presenceOf(b), b.id)
end

-- ------- handlers

local handlers = {}

handlers[Wire.HELLO] = function(self, client, msg)
  if client.ready then return end
  if Wire.int(msg.proto, 0, 9999) ~= Config.PROTOCOL then
    return self:refuseClient(client, ("This game speaks protocol %d; yours "
      .. "speaks %s."):format(Config.PROTOCOL, tostring(msg.proto)))
  end
  local name = Wire.name(msg.name)
  if not name then
    return self:refuseClient(client, "That trainer name can't be used here.")
  end
  -- the seat is claimed here, by someone who has identified themselves
  if self:isFull() then
    return self:refuseClient(client,
      ("This hub is full (%d players)."):format(self.limit))
  end

  client.name = name
  client.sprite = Wire.spriteId(msg.sprite) or Config.DEFAULT_SPRITE
  client.map = Wire.mapId(msg.map)
  client.x = Wire.int(msg.x, 0, 4096)
  client.y = Wire.int(msg.y, 0, 4096)
  client.facing = Wire.facing(msg.facing) or "down"
  client.ready = true
  self.players = self.players + 1

  local players = {}
  for id, other in pairs(self.clients) do
    if other.ready and id ~= client.id then
      players[#players + 1] = presenceOf(other)
    end
  end
  send(client, Wire.WELCOME, { id = client.id, players = players })
  self:broadcast(Wire.JOIN, { player = presenceOf(client) }, client.id)
end

handlers[Wire.MOVE] = function(self, client, msg)
  if not client.ready then return end
  local map = Wire.mapId(msg.map)
  local x, y = Wire.int(msg.x, 0, 4096), Wire.int(msg.y, 0, 4096)
  if map and x and y then
    client.map, client.x, client.y = map, x, y
  else
    -- no cell means "not in the world right now" (a battle, a menu): the
    -- player stays on the roster but stops being placeable
    client.map, client.x, client.y = nil, nil, nil
  end
  if Wire.facing(msg.facing) then client.facing = msg.facing end
  self:broadcast(Wire.MOVE, presenceOf(client), client.id)
end

handlers[Wire.CHAT] = function(self, client, msg)
  if not client.ready then return end
  local scope = Wire.SCOPES[msg.scope] and msg.scope or nil
  local text = Wire.text(msg.text, Config.MESSAGE_MAX)
  if not (scope and text) then return end

  if self.clock - client.lastChat < Config.CHAT_GATE then return end
  client.lastChat = self.clock

  local payload = { from = client.id, name = client.name,
                    scope = scope, text = text }

  if scope == "private" then
    local target = self.clients[Wire.id(msg.to) or ""]
    if target and target.ready then send(target, Wire.CHAT, payload) end
    return
  end

  if scope == "local" then
    if not client.map then return end
    for id, other in pairs(self.clients) do
      if id ~= client.id and other.ready and other.map == client.map then
        local distance = math.max(math.abs(other.x - client.x),
                                  math.abs(other.y - client.y))
        if distance <= Config.LOCAL_RADIUS then
          send(other, Wire.CHAT, payload)
        end
      end
    end
    return
  end

  self:broadcast(Wire.CHAT, payload, client.id)
end

handlers[Wire.REQUEST] = function(self, client, msg)
  if not client.ready or client.sessionId then return end
  local kind = Wire.KINDS[msg.kind] and msg.kind or nil
  local target = self.clients[Wire.id(msg.to) or ""]
  if not (kind and target and target.ready) or target.id == client.id then
    return
  end
  if target.sessionId then
    return send(client, Wire.DECLINE, { name = target.name, kind = kind })
  end
  client.pendingTo = target.id
  send(target, Wire.REQUEST,
    { from = client.id, name = client.name, kind = kind })
end

handlers[Wire.RESPOND] = function(self, client, msg)
  if not client.ready then return end
  local kind = Wire.KINDS[msg.kind] and msg.kind or nil
  local asker = self.clients[Wire.id(msg.to) or ""]
  if not (kind and asker and asker.ready) then return end

  -- only the player who was actually asked can answer, and only while the
  -- ask is still outstanding
  if asker.pendingTo ~= client.id then return end
  asker.pendingTo = nil

  if not msg.accept then
    return send(asker, Wire.DECLINE, { name = client.name, kind = kind })
  end
  if client.sessionId or asker.sessionId then
    return send(asker, Wire.DECLINE, { name = client.name, kind = kind })
  end
  self:startSession(asker, client, kind)
end

handlers[Wire.RELAY] = function(self, client, msg)
  if not client.ready or not client.sessionId then return end
  local peer = self:peerOf(client)
  if not peer then return end
  if Wire.id(msg.to) ~= peer.id then return end
  if not Wire.payloadOk(msg.payload) then return end
  -- The hub does not read the payload. It is the engine's own link
  -- vocabulary, and interpreting it here would couple this to a protocol
  -- the game already owns.
  send(peer, Wire.RELAY, { from = client.id, payload = msg.payload })
end

handlers[Wire.SESSION_LEAVE] = function(self, client)
  self:endSession(client, "peer_left")
end

handlers[Wire.PING] = function(self, client)
  send(client, Wire.PONG, {})
end

-- ------- entry points

function M:receive(client, msg)
  if not (client and self.clients[client.id]) then return end
  if type(msg) ~= "table" or type(msg.type) ~= "string" then return end
  local handler = handlers[msg.type]
  if handler then handler(self, client, msg) end
end

function M:update(dt)
  self.clock = self.clock + (dt or 0)

  -- Reap connections that never introduced themselves. Without this a peer
  -- can hold a slot indefinitely simply by saying nothing.
  local stale
  for _, client in pairs(self.clients) do
    if not client.ready
       and (self.clock - (client.since or 0)) > Config.HELLO_TIMEOUT then
      stale = stale or {}
      stale[#stale + 1] = client
    end
  end
  for _, client in ipairs(stale or {}) do
    if client.peer then
      client.peer:send({ type = Wire.ERROR, message = "Took too long to join." })
      client.peer:close()
    end
    self:drop(client)
  end
end

-- Tell everyone the game is over, then forget them. Called when the host
-- stops hosting or leaves: there is no host migration, so the honest thing
-- is to say so rather than leave clients talking to a dead listener.
function M:shutdown(message)
  for _, client in pairs(self.clients) do
    if client.peer then
      client.peer:send({
        type = Wire.ERROR,
        message = message or "The host ended the game.",
      })
      client.peer:close()
    end
  end
  self.clients, self.count, self.players, self.sessions = {}, 0, 0, {}
end

M.presenceOf = presenceOf

return M
